Handle part of missing VoiceOver accessibility in Shipping Labels flow

RafaelKayumov opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
Part of: WOOMOB-618

## Description
Addresses part of voice over issues listed in the document: 1nxDAV51DK_l8opZ2ZznqaA3FpfiYPoWZPd38HA4TTys-gdoc

- Grouped the collapsible products section on "Create Shipping Labels" screen
- Applied custom narration for the products section mentioned above
- Applied a hint about expanding/collapsing the product section
- Applied custom narration for the product items card
- Added grouping for shipments top tabs
- The "pencil" edit button narrates a hint that it navigates to shipments editing a.k.a "split shipments".
- Product item cards custom narration includes "number of items" value
- Omits empty dimensions in product item details.

## Testing steps

- Open a completed order with several shippable products/items
- Navigate to the "Create Shipping Label" flow
- Enable voice over

### Omitting empty product dimensions

- Find an order with products that miss dimensions.
- Navigate to the "Create Shipping Labels" flow
- Make sure the product card for that item doesn't display missing dimensions.

Before | After
--- | ---
<img width="458" height="137" alt="Снимок экрана 2025-07-15 в 13 43 54" src="https://github.com/user-attachments/assets/043d11de-e977-4c80-948b-919362e1843e" /> |  <img width="457" height="141" alt="Снимок экрана 2025-07-15 в 13 46 54" src="https://github.com/user-attachments/assets/08d2091d-0ad3-44e0-86c3-3032346bac55" />
<img width="459" height="145" alt="Снимок экрана 2025-07-15 в 13 44 06" src="https://github.com/user-attachments/assets/715d5426-e03f-4fbb-81c0-e1b3d9de7906" /> | <img width="457" height="146" alt="Снимок экрана 2025-07-15 в 13 47 03" src="https://github.com/user-attachments/assets/0213ba3e-0e5d-4871-95a9-f059a3ffd592" />

### Collapsed products section header

- Select the products section in collapsed state
- The whole section should be selected by a VoiceOver frame including the border
- Content should be narrated properly including the amount, weight and price
- The hint about expanding should be narrated

![Заказы 2](https://github.com/user-attachments/assets/92750a72-7914-4fa9-9174-1a12edc830f1)

### Expanded products section header

- Select the header
- The whole section should be selected by a VoiceOver frame. This time there is no border.
- The hint about collapsing should be narrated.

![Заказы 3](https://github.com/user-attachments/assets/d7b07ebe-cda0-452e-95c1-efe907ad6618)

### Product items cards in expanded products section

- Select a product item
- Each product item card should be selectable by the VoiceOver frame
- Card content should be narrated including the product name, details, weight and price. Cards with amount > 1 should also narrate the items count.

![Заказы 4](https://github.com/user-attachments/assets/4098a3f9-d0af-4394-9fc8-2a8c1582aff7)

### Shipments selection tabs

- Tap on a shipment tab in the top tab bar
- The whole tab content should be selected by the VoiceOver frame
- The tab content should be narrated including the tab selected state

![Заказы 5](https://github.com/user-attachments/assets/98c3b840-8d9c-4a07-b1d6-5cf19ac34be7)

### Edit shipments button

- Tap on a "Pencil" button in the shipments tab bar.
- The hint should narrate that it opens the shipment editing form.

![Заказы 6](https://github.com/user-attachments/assets/6f9a0b6b-f2c5-4d50-aee4-73d319b5256f)

## Testing information
I suggest to test on physical device with a left action button bound to VoiceOver toggling (for example iPhone 15 Pro).

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

RafaelKayumov committed:
Add products grouping and expand hint for voice over

RafaelKayumov committed:
Re-arrange accessibility and manage voice over for products section

RafaelKayumov committed:
Add manual accessibility value for product item row

RafaelKayumov committed:
Group tab elements in TopTabView tabs

RafaelKayumov committed:
Add accessibility hint for edit shipments button

RafaelKayumov committed:
Group elements for split shipment item row

RafaelKayumov committed:
Omit empty dimensions from details label

RafaelKayumov committed:
Update release notes

RafaelKayumov committed:
Update tests

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/ShipmentDetails/WooShippingShipmentDetailsViewModel.swift:
```diff
@@ -561,5 +573,11 @@ private extension WooShippingShipmentDetailsViewModel {
         )
+        static let itemsSummaryAccessibilityFormat = NSLocalizedString(
+            "shipping-labels.packages.items.summary.accessibility-label",
+            value: "%@ with a total weight of %@ and a total price of %@",
```
For localized strings it's recommended to use [positional placeholders](https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/localization.md#always-use-positional-placeholders) to help translators position the placeholders properly for different languages.

RafaelKayumov replied:
Done in [3d67589](https://github.com/woocommerce/woocommerce-ios/pull/15912/commits/3d675890a1d56c277323801afb667b537c6859d0)

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Items Section/ShippingItemRowAccessibility.swift:
```diff
@@ -0,0 +1,47 @@
+    ) -> String {
+        /// Covers item name and quantity if plural
+        let quantifiedItemFormattedString: String
+        if let quantityIntValue = Int(quantity), quantityIntValue > 1 {
```
Maybe we should explicitly announce the number of items even when it's 1 item, for clarity?

RafaelKayumov replied:
For now I don't feel like investing more time in it. The current narration looks good enough to me. We can get back to it later if needed.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Items Section/WooShippingItemRowViewModel.swift:
```diff
@@ -71,8 +71,23 @@ private extension ShippingLabelPackageItem {
+            return nil
+        }
+
+        return validDimensions.joined(separator: Constants.dimensionsFormatSeparator) + " " + unit
```
Nit: I'm unsure if this format works for every languages. Should we move it to a localized string?

RafaelKayumov replied:
I checked the localized formats before rearranging that. All formats and order looked the same. Would you suggest to introduce a format with 2 arguments where the 1st argument is dimensions(L x N x N) and the 2nd is the measure unit?

itsmeichigo replied:
Yes, that's what I meant, as I see the original code using that localized format.

itsmeichigo reviewed (commented):
Thank you for adding these improvements for voiceover 👍 

I spot a few issues:
- The fulfilled shipment tabs are announced like "Shipment 4, selected", maybe due to the check icon. It should be "Shipment 4, fulfilled" instead.
- The items on the split view screen can be further improved for accessibility:
  - For selectable parent items, there should be a hint saying that tapping on the item would expand or collapse the sub item list.
  - The checkboxes should have accessibility trait as buttons with value about selected state and hint about tapping to select the items.

I left some other nit-picking in the comments below.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Split Shipments/SelectableShipmentItemRow.swift:
```diff
@@ -22,40 +22,55 @@ struct SelectableShipmentItemRow: View {
+                VStack(alignment: .leading) {
+                    Text(viewModel.item.name)
+                        .bodyStyle()
+                    Text(viewModel.item.detailsLabel)
```
Should we check if the content is empty before rendering? That would avoid an empty line when there is no content:

<img width="354" height="147" alt="Screenshot 2025-07-16 at 10 03 48" src="https://github.com/user-attachments/assets/6e9f5c02-bd6e-48f9-a05c-67be2f42b930" />

RafaelKayumov replied:
In the current view structure the price view is aligned with the weight label since they are in the same horizontal stack. Skipping the empty value of "detailsLabel" and lifting up the weight will end up in price label moving up as well. I'm fine with checking the presence of the content and skipping the empty line. But that would mean re-arranging the layout here. Should I go for that? If yes - I'll work on that in a separate branch.

itsmeichigo replied:
Good point - I'm good with keeping this as-is then.

RafaelKayumov committed:
Apply fulfilled / unfulfilled accessibility value for shipment tabs

RafaelKayumov committed:
Index localized string parameters

RafaelKayumov commented:
> The fulfilled shipment tabs are announced like "Shipment 4, selected", maybe due to the check icon. It should be "Shipment 4, fulfilled" instead.
- Addressed in[f236459](https://github.com/woocommerce/woocommerce-ios/pull/15912/commits/f236459a1b192e7bab1bdcbefb853b6917061b91). I kept "selected" (which is tab selection state) and added fulfilled/unfulfilled for shipment tabs.

> The checkboxes should have accessibility trait as buttons with value about selected state and hint about tapping to select the items.
- I haven't yet grouped checkboxes completely. It's extracted as a separate element in child rows, but not yet extracted in parent rows. I'll address that in following PRs.

> For selectable parent items, there should be a hint saying that tapping on the item would expand or collapse the sub item list.
- I'll address that in the following PRs

itsmeichigo reviewed (approved):
Thanks for the updates!

RafaelKayumov committed:
Merge branch 'trunk' into WOOMOB-618-voice-over-issues-in-shipping-labels

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj
